backend/volunteer_hub/views.py

```diff
@@ -1,21 +1,133 @@
 from rest_framework.views import APIView
+from django.contrib.gis.db.models.functions import Distance
 from rest_framework.response import Response
 from rest_framework import status
 from rest_framework.permissions import AllowAny
-from .models import Volunteer
-from .serializers import VolunteerSerializer
+from django.contrib.postgres.search import SearchVector
+from django.db import models
+from django.db.models import Q
+from django.contrib.gis.geos import Point
+from django.contrib.gis.measure import D
+from .models import Volunteer, VolunteerTeam
+from .serializers import VolunteerSerializer, VolunteerTeamSerializer
 from drf_yasg.utils import swagger_auto_schema
 from drf_yasg import openapi
 
 
 class VolunteerListView(APIView):
-    permission_classes = [AllowAny]  # Allow any user to access this view
+    permission_classes = [AllowAny]
 
     @swagger_auto_schema(
-        operation_description="Retrieve a list of active volunteers",
+        operation_description="Retrieve a list of active volunteers with optional filtering and searching.",
+        manual_parameters=[
+            openapi.Parameter(
+                'thana', openapi.IN_QUERY, description="Filter by Thana ID", type=openapi.TYPE_INTEGER
+            ),
+            openapi.Parameter(
+                'district', openapi.IN_QUERY, description="Filter by District ID", type=openapi.TYPE_INTEGER
+            ),
+            openapi.Parameter(
+                'address', openapi.IN_QUERY, description="Full-text search on address", type=openapi.TYPE_STRING
+            ),
+            openapi.Parameter(
+                'latitude', openapi.IN_QUERY, description="Latitude for geospatial search", type=openapi.TYPE_NUMBER
+            ),
+            openapi.Parameter(
+                'longitude', openapi.IN_QUERY, description="Longitude for geospatial search", type=openapi.TYPE_NUMBER
+            ),
+        ],
         responses={200: VolunteerSerializer(many=True)},
     )
     def get(self, request, *args, **kwargs):
         volunteers = Volunteer.objects.filter(is_active=True)
+
+        # Filtering by Thana and District
+        thana_id = request.query_params.get('thana')
+        district_id = request.query_params.get('district')
+
+        if thana_id:
+            volunteers = volunteers.filter(thana_id=thana_id)
+        if district_id:
+            volunteers = volunteers.filter(district_id=district_id)
+
+        address_query = request.query_params.get('address')
+        if address_query:
+            volunteers = volunteers.filter(
+                address__icontains=address_query
+            ).distinct()
```

Comment:
Filtering with the address might be harmful to this scenario. Suppose you have 20 instances of Volunteer in the volunteers variable. Filtering with address query may exclude almost all instances If they do not match properly as there is a high chance of that.
But you see, those 20 volunteer instances were useful for us. So, I think you can give this another thought brother.